plasmapy/particles/_factory.py

```diff
@@ -9,15 +9,106 @@
 import astropy.units as u
 import contextlib
 
-from astropy.units.physical import electrical_charge, mass
-from numbers import Integral
-from typing import Union
+from astropy.constants import m_e
+from numbers import Integral, Real
+from typing import Any, Union
 
 from plasmapy.particles.exceptions import InvalidParticleError
 from plasmapy.particles.particle_class import CustomParticle, Particle
 from plasmapy.particles.particle_collections import ParticleList
 
 
+def _generate_error_message(args: tuple, kwargs: dict[str, Any]) -> str:
+    """Compose an error message for invalid particles."""
+
+    errmsg = "Unable to create a particle from: "
+
+    if args:
+        errmsg += repr(args)
+        if kwargs:
+            errmsg += " and "
+
+    if kwargs:
+        errmsg += repr(kwargs)
+
+    errmsg += (
+        ". For information on creating particles, see:"
+        "https://docs.plasmapy.org/en/stable/api/plasmapy.particles.ParticleLike.html"
+    )
+
+    return errmsg
+
+
+def _make_custom_particle_with_real_charge_number(*args, **kwargs):
+    """
+    Create a |CustomParticle| for mean or composite ions.
+
+    This function is intended to produce |CustomParticle| instances
+    provided a string representing an element or isotope without charge
+    information (e.g., ``"He"`` or ``"He-4"`` but not ``"He-4 2+"``)
+    along with a charge number that is a real number but not an integer.
+
+    Parameters
+    ----------
+    *args : (1,) tuple of str
+        A `tuple` containing a representation of an element or isotope,
+        without charge information.
+
+    *kwargs : dict of str to real numbers
+        Keyword arguments like those that can be passed to |Partile|,
+        except where the charge number ``Z`` is a real number but not
+        an integer.
```

Comment:
So, why can't the signature of this be along the lines of  `_make_custom...(s: str, Z: Optional[Union[Number, u.Quantity]], mass_numb: Optional[Number])`? If `args` is gonna be a 1-tuple, it just feels like an extra step, no?

Reply:
I think you're right.  The main thing we need to do with this is make sure that when this fails, it raises an exception that then gets handled. When arguments don't match the signature of a function, a `TypeError` gets raised, which is fortunately one of the ones that gets handled by `contextlib.suppress` further below.